### PR TITLE
Add nodegroup tags and DgElementArrayMemberBase class

### PR DESCRIPTION
**Pull request by nilsdeppe**
## Proposed changes

- These are currently the tags I know I'll need. There's another part I'm trying to figure out if I can avoid adding a tag.
- I might change `ElementCollection` to just hold `unique_ptr`s of the base. This is fairly likely since it helps decouple a lot of the code.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/ParallelAlgorithms/Events/ObserveTimeStep.hpp [hunk 1]
```diff
@@ -153,8 +153,8 @@ class ObserveTimeStep : public Event {
       "evolutions.";
 
   ObserveTimeStep() = default;
-  explicit ObserveTimeStep(const std::string& subfile_name,
-                           const bool output_time, const bool observe_per_core);
+  explicit ObserveTimeStep(const std::string& subfile_name, bool output_time,
```
Comment (kidder) [inline src/ParallelAlgorithms/Events/ObserveTimeStep.hpp]: remove explicit keyword

File: src/Parallel/ArrayCollection/Tags/ElementCollection.hpp [hunk 2]
```diff
@@ -0,0 +1,32 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <cstddef>
+#include <unordered_map>
+
+#include "DataStructures/DataBox/Tag.hpp"
+#include "Domain/Structure/ElementId.hpp"
```
Comment (kidder) [inline src/Parallel/ArrayCollection/Tags/ElementCollection.hpp]: forward declare

File: src/Parallel/ArrayCollection/Tags/NumberOfElementsTerminated.hpp [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+#pragma once
+
+#include <cstddef>
+
+#include "DataStructures/DataBox/Tag.hpp"
+
+namespace Parallel::Tags {
+/// \brief A counter of the number of elements that have marked themselves as
+/// terminated on the node.
```
Comment (kidder) [inline src/Parallel/ArrayCollection/Tags/NumberOfElementsTerminated.hpp]: terminated in what sense? the phase or program execution?

File: src/Parallel/ArrayCollection/Tags/ElementLocations.hpp [hunk 4]
```diff
@@ -0,0 +1,29 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <cstddef>
+#include <unordered_map>
+
+#include "DataStructures/DataBox/Tag.hpp"
+#include "Domain/Structure/ElementId.hpp"
```
Comment (kidder) [inline src/Parallel/ArrayCollection/Tags/ElementLocations.hpp]: forward declare

File: src/Parallel/ArrayCollection/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,4 @@
+# Distributed under the MIT License.
+# See LICENSE.txt for details.
+
```
Comment (kidder) [inline src/Parallel/ArrayCollection/CMakeLists.txt:3]: don't you need to add this subdirectory to the parent CMakeLists.txt?

File: src/Parallel/ArrayCollection/DgElementArrayMemberBase.hpp [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+#include <unordered_map>
+
+#include "Domain/Structure/ElementId.hpp"
+#include "Parallel/NodeLock.hpp"
+#include "Parallel/Phase.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/Serialization/CharmPupable.hpp"
+
+namespace Parallel {
+template <size_t Dim>
```
Comment (kidder) [inline src/Parallel/ArrayCollection/DgElementArrayMemberBase.hpp:33]: document the class

**Review by kidder — CHANGES_REQUESTED** [review]
you can squash the changes

File: src/Parallel/ArrayCollection/CMakeLists.txt (hunk 5 above)
Comment (nilsdeppe) [reply src/Parallel/ArrayCollection/CMakeLists.txt:3]: It is down below...

File: src/Parallel/CMakeLists.txt [hunk 7]
```diff
@@ -84,6 +84,7 @@ add_dependencies(
   )
 
 add_subdirectory(Algorithms)
+add_subdirectory(ArrayCollection)
```
Comment (nilsdeppe) [inline src/Parallel/CMakeLists.txt:88]: Here @kidder

File: src/Parallel/ArrayCollection/Tags/NumberOfElementsTerminated.hpp (hunk 3 above)
Comment (nilsdeppe) [reply src/Parallel/ArrayCollection/Tags/NumberOfElementsTerminated.hpp]: I guess if the program execution is terminated the value of this tag is ill-defined 🙃 but I'll be explicit

**Review by kidder — APPROVED** [review]

File: src/Parallel/ArrayCollection/Tags/ElementLocations.hpp [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+  using type = std::unordered_map<ElementId<Dim>, size_t>;
+};
+
+/// \brief The node (location) where different elements are.
+///
+/// This should be in the DgElementArrayMember's DataBox and should point to
+/// the one located in the nodegroup's DataBox.
+template <size_t Dim>
+struct ElementLocationsPointer : db::SimpleTag {
+  using type = std::unordered_map<ElementId<Dim>, size_t>*;
```
Comment (knelli2) [inline src/Parallel/ArrayCollection/Tags/ElementLocations.hpp:31]: How does this get set when pupping?

Comment (nilsdeppe) [reply src/Parallel/ArrayCollection/Tags/ElementLocations.hpp:31]: Currently there's no inter-node LB going on at all so pup can be ignored. Likely the parent nodegroup holding the `DgElementCollection` will need to set these in the children

**Pull request merged**